codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -243,66 +246,38 @@ public ChannelFuture writeHeaders(final ChannelHandlerContext ctx, final int str
                                 "Stream %d in unexpected state: %s", stream.id(), stream.state()));
                 }
             }
+            if (openStream) {
+                stream.open(endOfStream);
+            }
 
-            if (lastDataWrite != null && !endOfStream) {
-                throw new IllegalStateException(
-                        "Sending non-trailing headers after data has been sent for stream: "
-                                + streamId);
+            // Pass headers to the flow-controller so it can maintain their sequence relative to DATA frames.
+            ChannelFuture future = flowController().sendFlowControlledData(ctx, stream,
+                    new FlowControlledHeaders(ctx, stream, headers, streamDependency, weight,
+                            exclusive, padding, endOfStream,  promise),
+                    promise);
+            future.addListener(new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    if (!future.isSuccess()) {
+                        // The write failed, handle the error.
+                        lifecycleManager.onException(ctx, future.cause());
+                    } else if (endOfStream) {
+                        lifecycleManager.closeLocalSide(stream, ctx.voidPromise());
```

Comment:
On a related note I'm not sure the previous code was right, it was doing

lifecycleManager.closeLocalSide(stream, ctx.newPromise());

which if the channel was in a pending close state would cause it not to be notified as nothing ever completed the promise. Also not sure if it makes any sense to pass the future into the closeLocalSide method if it's always a ctx.succededFuture().